django_redshift_backend/base.py

```diff
@@ -553,6 +555,162 @@ class DatabaseCreation(BasePGDatabaseCreation):
     pass
 
 
+class DatabaseIntrospection(BasePGDatabaseIntrospection):
+    pass
+
+    def get_table_description(self, cursor, table_name):
+        """
+        Return a description of the table with the DB-API cursor.description
+        interface.
+        """
+        # Query the pg_catalog tables as cursor.description does not reliably
+        # return the nullable property and information_schema.columns does not
+        # contain details of materialized views.
+
+        # This function is based on the version from the Django postgres backend
+        # from before support for collations were introduced in Django 3.2
+        cursor.execute("""
+            SELECT
+                a.attname AS column_name,
+                NOT (a.attnotnull OR (t.typtype = 'd' AND t.typnotnull)) AS is_nullable,
+                pg_get_expr(ad.adbin, ad.adrelid) AS column_default
+            FROM pg_attribute a
+            LEFT JOIN pg_attrdef ad ON a.attrelid = ad.adrelid AND a.attnum = ad.adnum
+            JOIN pg_type t ON a.atttypid = t.oid
+            JOIN pg_class c ON a.attrelid = c.oid
+            JOIN pg_namespace n ON c.relnamespace = n.oid
+            WHERE c.relkind IN ('f', 'm', 'p', 'r', 'v')
+                AND c.relname = %s
+                AND n.nspname NOT IN ('pg_catalog', 'pg_toast')
+                AND pg_catalog.pg_table_is_visible(c.oid)
+        """, [table_name])
+        field_map = {
+            column_name: (is_nullable, column_default)
+            for (column_name, is_nullable, column_default) in cursor.fetchall()
+        }
+        cursor.execute(
+            "SELECT * FROM %s LIMIT 1" % self.connection.ops.quote_name(table_name)
+        )
+        return [
+            FieldInfo(
+                name=column.name,
+                type_code=column.type_code,
+                display_size=column.display_size,
+                internal_size=column.internal_size,
+                precision=column.precision,
+                scale=column.scale,
+                null_ok=field_map[column.name][0],
+                default=field_map[column.name][1],
+                collation=None,  # Redshift doesn't support user-defined collation
+                # https://docs.aws.amazon.com/redshift/latest/dg/c_collation_sequences.html
+            )
+            for column in cursor.description
+        ]
+
+    def get_constraints(self, cursor, table_name):
+        """
+        Retrieve any constraints or keys (unique, pk, fk, check, index) across
+        one or more columns. Also retrieve the definition of expression-based
+        indexes.
+        """
+        # Based on code from Django 3.2
```

Comment:
nits: I'll add a URL to original django source after merging:

```suggestion
        # Based on code from Django 3.2
        # https://github.com/django/django/blob/3.2.12/django/db/backends/
        # postgresql/introspection.py#L148-L182
```